plugins/org.eclipse.elk.core/src/org/eclipse/elk/core/RecursiveGraphLayoutEngine.java

```diff
@@ -194,8 +200,151 @@ protected List<ElkEdge> layoutRecursively(final ElkNode layoutNode, final TestCo
                 }
 
             } else {
-                // Layout each compound node contained in this node separately
                 nodeCount = layoutNode.getChildren().size();
+                // If performing topdown layout, layout this node first, then compute scale factors and layout children
+                // recursively
+                if (layoutNode.getProperty(CoreOptions.TOPDOWN_LAYOUT)) {
+                                                            
+                    IElkProgressMonitor topdownLayoutMonitor = progressMonitor.subTask(1);
+                    topdownLayoutMonitor.begin("Topdown Layout", 1);
+                    
+                    // if we are currently in a region and about to produce a child layout,
+                    // then we need to step through the states and set the sizes of the states 
+                    // only if their layout is topdownpacking
+                    if (layoutNode.getProperty(CoreOptions.TOPDOWN_NODE_TYPE).equals(TopdownNodeTypes.HIERARCHICAL_NODE)
+                            || layoutNode.getProperty(CoreOptions.TOPDOWN_NODE_TYPE).equals(TopdownNodeTypes.ROOT_NODE)) {
+                        
+                        for (ElkNode childNode : layoutNode.getChildren()) {
+                            // check if child has children and whether it will be laid out with topdownpacking
+                            // if yes its size needs to be pre-computed before computing the layout
+                            if (childNode.getChildren().size() > 0
+                                    && childNode.getProperty(CoreOptions.ALGORITHM).equals("org.eclipse.elk.topdownpacking")) {
+                                
+                                // get relevant properties
+                                ElkPadding padding = childNode.getProperty(CoreOptions.PADDING);
+                                double nodeNodeSpacing = childNode.getProperty(CoreOptions.SPACING_NODE_NODE);
+                                
+                                double hierarchicalNodeWidth 
+                                    = childNode.getProperty(CoreOptions.TOPDOWN_HIERARCHICAL_NODE_WIDTH);
+                                double hierarchicalNodeAspectRatio 
+                                    = childNode.getProperty(CoreOptions.TOPDOWN_HIERARCHICAL_NODE_ASPECT_RATIO);
+                                
+                                int numberOfChildren = childNode.getChildren().size();
+                                int cols = (int) Math.ceil(Math.sqrt(numberOfChildren));
+                                double requiredWidth = cols * hierarchicalNodeWidth 
+                                        + padding.left + padding.right + (cols - 1) * nodeNodeSpacing; 
+                                int rows;
+                                if (numberOfChildren > cols * cols - cols) {
```

Comment:
I read that before, looks like duplicate code.

Reply:
Yes this is a dependency to the node placer of the topdownpacking algorithm, this needs to be extracted